EZP-24704: Solr: Implement native result extractor

pspanja opened the pull request:
> ##### This PR resolves https://jira.ez.no/browse/EZP-24704
> ##### Review together with https://github.com/ezsystems/ezpublish-kernel/pull/1372

This implements *native* result extractor, which extracts search result from data returned by Solr backend.

`NativeResultExtractor` replaces `LoadingResultExtractor`, which did the "extraction" by loading from persistence by ID returned from the Solr backend.

The motive is completely avoiding hitting of persistence when using `SearchService::findContentInfo()`.

pspanja commented:
Ready for review, ping @andrerom @glye

andrerom commented on lib/eZ/Publish/Core/Search/Solr/ResultExtractor/NativeResultExtractor.php:
```diff
@@ -0,0 +1,102 @@
+     *
+     * @return \eZ\Publish\SPI\Persistence\Content\ContentInfo
+     */
+    protected function extractContentHit($hit)
```
`extractContentInfoFromHit`

pspanja replied:
Updated and rebased.

andrerom commented on lib/eZ/Publish/Core/Search/Solr/ResultExtractor/NativeResultExtractor.php:
```diff
@@ -0,0 +1,102 @@
+     *
+     * @return \eZ\Publish\SPI\Persistence\Content\Location
+     */
+    protected function extractLocationHit($hit)
```
`extractLocationFromHit`

pspanja replied:
Updated and rebased.

andrerom commented:
+1

pspanja commented:
> do we need to cast stuff here maybe? considering the change for weak comparison for kernel test

This implementation should work for any storage, but maybe we can cast for now to avoid breaks on `5.4`. We do hint IDs as `int|string`, so one should not rely on them being `int` alone.

andrerom commented:
> This implementation should work for any storage, 

ok, ideally id's are either hash or ints, not hash, ints and numerical strings. But unless we index information about this, or abstract data mappers so search engine reuses data mappers from storage engine or something, then I don't see a good way to deal with this cleanly.

pspanja commented:
Yeah, you're right. But let's get back at that later :)
Now added casts and rebased, also fixed version number not being indexed as integer in https://github.com/ezsystems/ezplatform-solr-search-engine/commit/7d18cbb7412ee0cffc36408507f251778ffefab4.

andrerom commented on lib/eZ/Publish/Core/Search/Solr/ResultExtractor/NativeResultExtractor.php:
```diff
@@ -0,0 +1,102 @@
+                'ownerId' => (int)$hit->owner_id,
+                'modificationDate' => strtotime($hit->modified_dt),
+                'publicationDate' => strtotime($hit->published_dt),
+                'alwaysAvailable' => $hit->always_available_b,
```
(bool)?

pspanja replied:
It comes as a bool since it is stored as such and delivered as JSON, which supports PHP's scalar types.
So no conversion needed other than for IDs.

andrerom replied:
ok, great.

andrerom commented on lib/eZ/Publish/Core/Search/Solr/ResultExtractor/NativeResultExtractor.php:
```diff
@@ -0,0 +1,102 @@
+                'name' => $hit->name_s,
+                'contentTypeId' => (int)$hit->type_id,
+                'sectionId' => (int)$hit->section_id,
+                'currentVersionNo' => $hit->version_no_i,
```
(int)

andrerom commented:
Ok, well safer for now.
And as types are probably ok now, +1
